### PR TITLE
Build image out of Dockerfile, run it and check that the container is up

### LINKED ISSUE #179: Build an image out of DockerFile with 'docker build' and run it
After having a DockerFile #176 , we wish to make an image out of it.

- Think (or lets discuss) where is the appropriate place in our code ('bootstrap.sh' ?) to put the build command.
- Add to appropriate place in a code, a command to build an image using the existing Dockerfile and run the container of of it.
- Add a test to tests/test_sanity.py checking that the image exists.

Acceptance Criteria:
- When deploying, an image is available and the container is up and running

### PR DESCRIPTION
**Pull request by IlanZuckerman**
fix #179 

Adding to Vagrant file the command to build the image and running it
with a name 'beyond/web_app'.
Adding 'docker_runner.sh' for running the docker container inside github actions (for CI tests)
Adding a new CI python test to check whether the 'beyond/web_app' container is
running.
Adding required package for the new test.

### CODE REVIEW

**Comment by lioramilbaum** [comment]
@IlanZuckerman You might want to move this PR to draft, as long as it is WIP

**Comment by IlanZuckerman** [comment]
Liora, no matter what i did (and i tried different approaches), i could not make the newly added test pass (it is triggered by github actions).
It almost looks like inside github actions, the container wont run.
In the Vagrant flow, however all is good, and the container is running as expected...

**Comment by IlanZuckerman** [comment]
> @IlanZuckerman You might want to move this PR to draft, as long as it is WIP

Actually, I wanted to hear your opinion. Since I'm pretty much out of ideas...

**Comment by ipolonsk** [comment]
Hey @IlanZuckerman,
I tried to test the changes you did to understand why the CI is not passing. (added the changes to my local repo and ran an VM at the stage env).
two small things I noticed:
1. After the VM was created I ssh into the server to check if the container is really running, after running 'docker ps'
the container status was "Restarting (0) 59 seconds ago" for the all-time. not sure if it's my mistake of copying the changes or the real status after the deployment. 
2. The test you did with docker using python, I think it tests the name incorrectly. I tried playing with it:
>>> client.containers.list()
[<Container: b120852aef>]
the list func returns a list of objects, no the name of the containers, so maybe you checking if the name is in that list and it not actioning as we wanted it to do. 
if we want to get the name of the container and more data we need to go through :
>>> container = client.containers.get('b120852aef')
>>> container.name
'beyond-web_app'
>>> container.status
'restarting'
3. we need to check no only if the name is in that list and also the status of it.

File: docker_runner.sh
Comment (lioramilbaum) [inline docker_runner.sh:10]: Is there a reason not to use the vagrant docker provisioner to run the container?

File: docker_runner.sh
Comment (lioramilbaum) [inline docker_runner.sh:9]: The build can be performed by GitHub Actions directly. Bash script is an overkill.

File: tests/test_sanity.py
Comment (lioramilbaum) [inline tests/test_sanity.py:35]: I think that inspec is a better suited technology for testing infrastructure.

**Review by lioramilbaum — CHANGES_REQUESTED** [review]

File: docker_runner.sh
Comment (IlanZuckerman) [reply docker_runner.sh:10]: Yes, because then the docker container would be running only when we run the Vagrant up.
On the CI flow it wont run if i would use the Vagrant docker provisioner.
I wanted to use the container both, in CI and regular Vagrant flow.
Thats why i took it out to external shell script, so i could invoke it both, from the Vagrant file AND the main.yml (from the github flow).

File: docker_runner.sh
Comment (IlanZuckerman) [reply docker_runner.sh:9]: See above comment where i explain why I choose this approach.

File: tests/test_sanity.py
Comment (IlanZuckerman) [reply tests/test_sanity.py:35]: Currently there is no inspec in our code, and i wanted to make a test for running container.
python 'docker' library is a convenient way of using docker API.

File: docker_runner.sh
Comment (lioramilbaum) [reply docker_runner.sh:10]: This is a valid point. docker-compose is a better suited technology then bash.

File: tests/test_sanity.py
Comment (lioramilbaum) [reply tests/test_sanity.py:35]: Please elaborate why it is convenient. Stating that one approach is more convenient than another is not something I or anyone else for that matter can learn from.
You can also defer the migration to inspec in a future task. I would accept the current solution as a "quick and dirty" one.

Comment (IlanZuckerman) [reply tests/test_sanity.py:35]: Alright, lets consider this as a temporary solution.
Integrating rspec will require some effort, and learning curve. While doing that, this patch will be thrown far back to the backlog.

File: docker_runner.sh
Comment (IlanZuckerman) [reply docker_runner.sh:10]: I am not very familiar with docker-compose, and how it can replace invocation of this shell script from two different places. But i will try to figure this out...

File: tests/test_sanity.py
Comment (lioramilbaum) [reply tests/test_sanity.py:35]: I am not following. We have decided that the pytest approach is accepted for now. The inspec will be implemented right after this one is merged.

Comment (IlanZuckerman) [reply tests/test_sanity.py:35]: Ok, thats what i was saying. Thats totally fine with me.
You were saying that inspec is a better suited here. I believe this is true.
But for sake of this patch, there is no argue IMHO.

Comment (lioramilbaum) [reply tests/test_sanity.py:35]: Please file an issue for the inspec effort. Start working on it after this PR is merged.

Comment (IlanZuckerman) [reply tests/test_sanity.py:35]: Created issue:
https://github.com/beyond-io/beyond/issues/190

**Review by ipolonsk — DISMISSED** [review]
LGTM

File: docker_runner.sh
Comment (IlanZuckerman) [reply docker_runner.sh:10]: I read a bit about docker-compose, to get a clue about what it is and what it is used for.
But still i dont understand how this can help us with the current duplication we have in invocation of 'docker run' (1- in docker_runner.sh -> for the 'Vagrant up' flow    2- in main.yml  -> for the CI flow.
We still need to invoke it from two sources if we wish to both - test in in the CI AND use it with the Vagrant flow.

I think for now we should review it again and consider merging it, since this task didnt define the usage of docker-compose in the first place. So I implemented it on the best side, but as i understood to make this possible.

In the future we could consider converting the flow to use docker compose.

**Review by lioramilbaum — APPROVED** [review]

**Comment by lioramilbaum** [comment]
Please rebase the PR

**Comment by IlanZuckerman** [comment]
> Please rebase the PR

Done

**Review by ipolonsk — APPROVED** [review]
LGTM

**Review by lioramilbaum — APPROVED** [review]

**Pull request merged**
